homeassistant/components/sensor/github.py

```diff
@@ -0,0 +1,335 @@
+        self._open_pull_requests = None
+        self._last_commit = None
+        self._latest_release = None
+        self._github_data = github_data
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._available
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            ATTR_PATH: self._repository_path,
+            ATTR_NAME: self._name,
+            ATTR_STARGAZERS: self._stargazers,
+            ATTR_TOPICS: self._topics,
+            ATTR_BRANCHES: self._branches,
+            ATTR_OPEN_ISSUES: self._open_issues,
+            ATTR_OPEN_PULL_REQUESTS: self._open_pull_requests,
+            ATTR_LAST_COMMIT: self._last_commit,
+            ATTR_LATEST_RELEASE: self._latest_release
+        }
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend."""
+        return 'mdi:github-circle'
+
+    def update(self):
+        """Collect updated data from GitHub API."""
+        self._github_data.update()
+
+        self._state = self._github_data.stargazers
+        self._repository_path = self._github_data.repository_path
+        self._name = self._github_data.name
+        self._stargazers = self._github_data.stargazers
+        self._topics = self._github_data.topics
+        self._branches = self._github_data.branches
+        self._open_issues = self._github_data.open_issues
+        self._open_pull_requests = self._github_data.open_pull_requests
+        self._last_commit = self._github_data.last_commit
+        self._latest_release = self._github_data.latest_release
+        self._available = self._github_data.available
+
+
+class GitHubData():
+    """GitHub Data object."""
+
+    def __init__(self, interval, repository, access_token=None, username=None,
+                 password=None, server_url=None):
+        """Set up GitHub."""
+        import github
+
+        self._github = github
+
+        self.setup_error = False
+
+        try:
+            if server_url is not None:
+                server_url += "/api/v3"
+                if access_token is not None:
+                    self._github_obj = github.Github(
+                        access_token, base_url=server_url)
+                elif username is not None and password is not None:
+                    self._github_obj = github.Github(
+                        username, password, base_url=server_url)
+            else:
+                if access_token is not None:
+                    self._github_obj = github.Github(access_token)
+                elif username is not None and password is not None:
+                    self._github_obj = github.Github(username, password)
```

Comment:
That's some slightly complex conditionals. I think it could be improved by constructing a `kwargs` and using a single `github.Github()` call.

Reply:
Not sure this is possible with this library. I have reduced the amount of logic by removing username/password so I think this should be OK